brainscore_language/utils/transformations.py

```diff
@@ -0,0 +1,248 @@
+        score = apply_aggregate(self.aggregate, score)
+        return score
+
+    def _run_pipe(self, *args, apply, **kwargs):
+        generator = self.pipe(*args, **kwargs)
+        for vals in generator:
+            y = apply(*vals)
+            done = generator.send(y)
+            if done:
+                break
+        result = next(generator)
+        return result
+
+    def pipe(self, *args, **kwargs):
+        raise NotImplementedError()
+
+    def _get_result(self, *args, done):
+        """
+        Yields the `*args` for further processing by coroutines
+        and waits for the result to be sent back.
+        :param args: transformed values
+        :param bool done: whether this is the last transformation and the next `yield` is the combined result
+        :return: the result from processing by the coroutine
+        """
+        result = yield args  # yield the values to coroutine
+        yield done  # wait for coroutine to send back similarity and inform whether result is ready to be returned
+        return result
+
+    def aggregate(self, score: Score) -> Score:
+        return Score(score)
+
+
+class Split:
+    class Defaults:
+        splits = 10
+        train_size = .9
+        split_coord = 'stimulus_id'
+        stratification_coord = None
+        unique_split_values = False
+        random_state = 1
+
+    def __init__(self,
+                 splits=Defaults.splits, train_size=None, test_size=None,
+                 split_coord=Defaults.split_coord, stratification_coord=Defaults.stratification_coord, kfold=False,
+                 unique_split_values=Defaults.unique_split_values, random_state=Defaults.random_state):
+        super().__init__()
+        if train_size is None and test_size is None:
+            train_size = self.Defaults.train_size
+        if kfold:
+            assert (train_size is None or train_size == self.Defaults.train_size) and test_size is None
+            if stratification_coord:
+                self._split = StratifiedKFold(n_splits=splits, shuffle=True, random_state=random_state)
+            else:
+                self._split = KFold(n_splits=splits, shuffle=True, random_state=random_state)
+        else:
+            if stratification_coord:
+                self._split = StratifiedShuffleSplit(
+                    n_splits=splits, train_size=train_size, test_size=test_size, random_state=random_state)
+            else:
+                self._split = ShuffleSplit(
+                    n_splits=splits, train_size=train_size, test_size=test_size, random_state=random_state)
+        self._split_coord = split_coord
+        self._stratification_coord = stratification_coord
+        self._unique_split_values = unique_split_values
+
+        self._logger = logging.getLogger(fullname(self))
+
+    @property
+    def do_stratify(self):
+        return bool(self._stratification_coord)
+
+    def build_splits(self, assembly):
```

Comment:
flagging that some of this code did not work in the brainio / neuralNLP repo when Carina and I performed checks on stratification. it had no effect. we could not debug it

Reply:
do you have a unit test showcasing that?

Reply:
no, sorry. only many 'manual' print statements (we slacked about this too some months back, can find the thread if relevant).